### PR TITLE
Solved bug when trying to find node info in namespaces

### LINKED ISSUE #181: [ros2node] info requires fully qualified node name
## Bug report

**Required Info:**

- Operating System:
  - Ubuntu bionic arm64
- Installation type:
  -  packaging job 
- Version or commit hash:
  - https://ci.ros2.org/view/packaging/job/packaging_linux-aarch64/649/
- DDS implementation:
  - Fast-RTPS
- Client library (if applicable):
  - N/A

#### Steps to reproduce issue

1. Launch a hidden node
   ```
   ros2 daemon start
   ```
1. Try to get info on the node using a fully qualified nod ename
   ```
   ros2 node info /_ros2cli_node_daemon_0
   ```
1. Try to get info on the node using a non fully qualified node name
   ```
   ros2 node info _ros2cli_node_daemon_0
   ```

#### Expected behavior

Both commands would list the info about the node

#### Actual behavior

The non-fully qualified version can't find the node.

```bash 
$ ros2 node info /_ros2cli_node_daemon_0
/_ros2cli_node_daemon_0
  Subscribers:

  Publishers:
    /parameter_events: rcl_interfaces/ParameterEvent
  Services:
    /_ros2cli_node_daemon_0/describe_parameters: rcl_interfaces/DescribeParameters
    /_ros2cli_node_daemon_0/get_parameter_types: rcl_interfaces/GetParameterTypes
    /_ros2cli_node_daemon_0/get_parameters: rcl_interfaces/GetParameters
    /_ros2cli_node_daemon_0/list_parameters: rcl_interfaces/ListParameters
    /_ros2cli_node_daemon_0/set_parameters: rcl_interfaces/SetParameters
    /_ros2cli_node_daemon_0/set_parameters_atomically: rcl_interfaces/SetParametersAtomically
$ ros2 node info _ros2cli_node_daemon_0
Unable to find node '_ros2cli_node_daemon_0'
```

### LINKED ISSUE #201: [ros2node] info fails to find nodes in namespaces
## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 18.04
- Installation type:
  - Source
- Version or commit hash:
  - ros2cli 7a65e1ecf02c69bf3d9624d49b17cde57f7ed16a
  - rmw_fastrtps d8f1959a785ad770f96623d257b8a4ae1c74c938
- DDS implementation:
  - Fast-RTPS
- Client library (if applicable):
  - N/A

#### Steps to reproduce issue

1. Save script as `namespace.launch.py`
```
#!/usr/bin/env python3

from launch import LaunchDescription

from launch_ros.actions import Node


def generate_launch_description():
    talker1 = Node(package='demo_nodes_cpp', node_executable='talker')
    talker2 = Node(package='demo_nodes_cpp', node_executable='talker', node_namespace='ns')

    return LaunchDescription([talker1, talker2])
```

2. Run `ros2 launch namespace.launch.py`

3. Run `ros2 node info /talker`

4. Run `ros2 node info /ns/talker`

#### Expected behavior
Information is printed for both nodes.

#### Actual behavior
Node inside a namespace throws
```
[ERROR] [rmw_fastrtps_shared_cpp]: Unable to find GUID for node: talker
Failed to get_subscriber_names_and_types: Unable to find GUID for node , at /home/jacob/ros2_core_ws/src/ros2/rmw_fastrtps/rmw_fastrtps_shared_cpp/src/rmw_node_info_and_types.cpp:85
```

### PR DESCRIPTION
**Pull request by ivanpauno**
Fixes #201 
Fixes #181 
Initial '/' was missing when the node was inside a namespace.
Now, non fully qualified names are also accepted.
Added unit test.

### CODE REVIEW

**Comment by ivanpauno** [comment]
Maybe, I should also correct https://github.com/ros2/ros2cli/issues/181 here

File: ros2node/ros2node/api/__init__.py [hunk 1]
```diff
@@ -37,8 +37,8 @@ def parse_node_name(full_node_name):
     namespace = '/'
     if len(tokens) > 1:
         node_name = tokens[-1]
-        namespace = '/'.join(tokens[:-1])
-    return NodeName(node_name, namespace, full_node_name)
+        namespace = '/' + '/'.join(tokens[:-1])
+    return NodeName(node_name, namespace, '/' + full_node_name)
```
Comment (dirk-thomas) [inline ros2node/ros2node/api/__init__.py]: The terminology seems to be all mixed in this function. The parameter `full_node_name` must explicitly not have a leading slash. But the returned value of `NodeName.full_name` explicitly does.

While not part of this patch the condition `1 > len(tokens)` can never be true. Also some callers explicitly remove the leading slash before calling the function.

Comment (dirk-thomas) [reply ros2node/ros2node/api/__init__.py]: > * Change callers to not remove the leading slash.
> * Also fix #181 here, adding a leading slash to non fully qualified names (inside parse_node_name).
> * Change "parse_node_name(full_node_name)" to "parse_node_name(node_name)" ...

:+1: 

> ... and node_name to "bare_node_name" or something similar.

Maybe `node_basename` to match terminology for paths?

**Comment by ivanpauno** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6408/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2791/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5266/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6240/)

**Comment by ivanpauno** [comment]
@dirk-thomas The failing test is unrelated, so I will merge if you think it is ok.

File: ros2node/ros2node/api/__init__.py [hunk 2]
```diff
@@ -29,16 +29,18 @@ def get_absolute_node_name(node_name):
...
 
 
-def parse_node_name(full_node_name):
-    tokens = full_node_name.split('/')
-    if 1 > len(tokens):
-        raise RuntimeError('Invalid node name: ' + full_node_name)
-    node_name = full_node_name
-    namespace = '/'
+def parse_node_name(node_name):
+    if (node_name[0] == '/'):
```
Comment (dirk-thomas) [inline ros2node/ros2node/api/__init__.py]: Nitpick: no need for parenthesis around conditions.

File: ros2node/ros2node/api/__init__.py [hunk 3]
```diff
@@ -29,16 +29,18 @@ def get_absolute_node_name(node_name):
...
+        node_name = node_name[1:]
+    tokens = node_name.split('/')
```
Comment (dirk-thomas) [inline ros2node/ros2node/api/__init__.py]: Since the following code only uses two parts of this I would suggest splitting explicitly for that: `rsplit('/', 1)`

With this change it might be better to prepend `/` on demand to the parameter rather then removing it if present. Just an idea.

**Comment by ivanpauno** [comment]
New CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6413/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2796/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5271/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6245/)

**Comment by dirk-thomas** [comment]
I simplified the logic a bit in c5c72d41f5378e7ceaaf442aded4cc2e5f47ec55. That unfortunately broke the DCO check. Feel free to override it (or squash before merging to make it pass again).

**Comment by ivanpauno** [comment]
Last CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6422/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2804/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5279/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6253/)

**Comment by ivanpauno** [comment]
Merging, as CI failures are unrelated

**Comment by Karsten1987** [comment]
Could it be that this issue still persists? 

```
 ➭ ros2 node info /twist_publisher
/twist_publisher
  Subscribers:
    /parameter_events: rcl_interfaces/ParameterEvent
  Publishers:
    /cmd_vel: geometry_msgs/Twist
    /parameter_events: rcl_interfaces/ParameterEvent
    /rosout: rcl_interfaces/Log
  Services:
    /twist_publisher/describe_parameters: rcl_interfaces/DescribeParameters
    /twist_publisher/get_parameter_types: rcl_interfaces/GetParameterTypes
    /twist_publisher/get_parameters: rcl_interfaces/GetParameters
    /twist_publisher/list_parameters: rcl_interfaces/ListParameters
    /twist_publisher/set_parameters: rcl_interfaces/SetParameters
    /twist_publisher/set_parameters_atomically: rcl_interfaces/SetParametersAtomically
```
when invoked without `/` I'll get:
```
 ➭ ros2 node info twist_publisher
Unable to find node 'twist_publisher'
```
